src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SoftReference`1.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace CSharpSyntaxGenerator
+{
+    /// <summary>
+    /// This class behaves like <see cref="WeakReference{T}"/>, with the added constraint that the referenced object
+    /// will not be reclaimed prior to the next Generation 2 garbage collection operation.
+    /// </summary>
+    /// <remarks>
+    /// The Generation 2 constraint only applies to the current target of the soft reference. If the target is replaced
+    /// by calling <see cref="SetTarget(T)"/>, the previous target may be immediately eligible for garbage collection.
+    /// </remarks>
+    /// <typeparam name="T">The type of object referenced.</typeparam>
+    internal sealed class SoftReference<T>
+        where T : class?
+    {
+        private readonly WeakReference<T?> _weakReference = new(null);
+        private T? _strongReference;
+
+        public SoftReference(T? target)
+        {
+            if (target is not null)
+                SetTarget(target);
+
+            Gen2GcCallback.Register(
+                static self => ((SoftReference<T>)self).OnGen2GC(),
+                this);
+        }
+
+        public void SetTarget(T target)
```

Comment:
is it desirable to make this typ emutable?  as opposed to just having the client just construct a new instance?

Reply:
is the presumption here that NRT wil ensure that peopel don't try to SetTarget with null?  should we explicitly try to stop people from doing that?

Reply:
My goal was to create a type that looks and feels like `WeakReference<T>` but tends to hold on to objects for a bit longer.